File: android/app/src/main/java/com/foodsense/android/services/AuthManager.kt
package com.foodsense.android.services

import android.util.Log
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.tasks.await

class AuthManager {
    private val auth: FirebaseAuth? = try {
        FirebaseAuth.getInstance()
    } catch (e: Exception) {
        Log.w("AuthManager", "Firebase Auth unavailable: ${e.message}")
        null
    }

    // If Firebase is unavailable, treat as signed in so user can use the app
    private val _isSignedIn = MutableStateFlow(auth?.currentUser != null || auth == null)
    val isSignedIn: StateFlow<Boolean> = _isSignedIn

    private val _currentUser = MutableStateFlow(auth?.currentUser)
    val currentUser: StateFlow<FirebaseUser?> = _currentUser

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error

    init {
        auth?.addAuthStateListener { firebaseAuth ->
            _currentUser.value = firebaseAuth.currentUser
            _isSignedIn.value = firebaseAuth.currentUser != null
        }
    }

    suspend fun signInAnonymously() {
        if (auth == null) {
            // Firebase not configured — skip auth and let user through
            _isSignedIn.value = true
            return
        }
        try {
            auth.signInAnonymously().await()
            _error.value = null
        } catch (e: Exception) {
            Log.w("AuthManager", "Anonymous sign-in failed: ${e.message}")
            _error.value = null
            // Let the user through anyway so the app is usable without Firebase
            _isSignedIn.value = true
        }
    }

    suspend fun signInWithEmail(email: String, password: String) {
        if (auth == null) {
            _error.value = "Authentication service unavailable. Set up Firebase to enable sign-in."
            return
        }
        try {
            auth.signInWithEmailAndPassword(email, password).await()
            _error.value = null
        } catch (e: Exception) {
            _error.value = e.localizedMessage ?: "Sign-in failed"
        }
    }

    suspend fun createAccount(email: String, password: String) {
        if (auth == null) {
            _error.value = "Authentication service unavailable. Set up Firebase to enable accounts."
            return
        }
        try {
            auth.createUserWithEmailAndPassword(email, password).await()
            _error.value = null
        } catch (e: Exception) {
            _error.value = e.localizedMessage ?: "Account creation failed"
        }
    }

    fun signOut() {
        auth?.signOut()
        if (auth == null) _isSignedIn.value = false
    }

    suspend fun deleteAccount() {
        try {
            auth?.currentUser?.delete()?.await()
        } catch (e: Exception) {
            Log.w("AuthManager", "Delete account failed: ${e.message}")
        }
    }
}

File: FoodDetectionApp/AuthManager.swift
import Foundation
import FirebaseAuth
import Combine

class AuthManager: ObservableObject {
    static let shared = AuthManager()

    @Published var isSignedIn = false
    @Published var currentUser: User?
    @Published var isLoading = true
    @Published var errorMessage: String?

    private var firebaseAvailable: Bool
    private var authStateHandle: AuthStateDidChangeListenerHandle?

    init() {
        // Check if Firebase is configured before accessing Auth
        let hasFirebase = Bundle.main.path(forResource: "GoogleService-Info", ofType: "plist") != nil
        self.firebaseAvailable = hasFirebase

        if hasFirebase {
            authStateHandle = Auth.auth().addStateDidChangeListener { [weak self] _, user in
                DispatchQueue.main.async {
                    self?.currentUser = user
                    self?.isSignedIn = user != nil
                    self?.isLoading = false
                }
            }
        } else {
            // No Firebase — treat as signed in so user can use the app
            self.isSignedIn = true
            self.isLoading = false
            print("[FoodSense] Firebase Auth unavailable — running without authentication")
        }
    }

    func signInAnonymously() async throws {
        guard firebaseAvailable else {
            await MainActor.run {
                self.isSignedIn = true
                self.isLoading = false
            }
            return
        }
        do {
            let result = try await Auth.auth().signInAnonymously()
            await MainActor.run {
                self.currentUser = result.user
                self.isSignedIn = true
                self.errorMessage = nil
            }
        } catch {
            print("[FoodSense] Anonymous sign-in failed: \(error.localizedDescription)")
            // Let user through anyway
            await MainActor.run {
                self.isSignedIn = true
                self.isLoading = false
                self.errorMessage = nil
            }
        }
    }

    func signInWithEmail(email: String, password: String) async throws {
        guard firebaseAvailable else {
            await MainActor.run {
                self.errorMessage = "Authentication unavailable. Set up Firebase to enable sign-in."
            }
            return
        }
        let result = try await Auth.auth().signIn(withEmail: email, password: password)
        await MainActor.run {
            self.currentUser = result.user
            self.isSignedIn = true
        }
    }

    func createAccount(email: String, password: String) async throws {
        guard firebaseAvailable else {
            await MainActor.run {
                self.errorMessage = "Authentication unavailable. Set up Firebase to enable accounts."
            }
            return
        }
        let result = try await Auth.auth().createUser(withEmail: email, password: password)
        await MainActor.run {
            self.currentUser = result.user
            self.isSignedIn = true
        }
    }

    func signOut() throws {
        if firebaseAvailable {
            try Auth.auth().signOut()
        }
        self.currentUser = nil
        self.isSignedIn = false
    }

    func deleteAccount() async throws {
        guard firebaseAvailable, let user = Auth.auth().currentUser else { return }
        try await user.delete()
        await MainActor.run {
            self.currentUser = nil
            self.isSignedIn = false
        }
    }
}
